### ISSUE TITLE
Publishing to SNS using Topic from external service API response

### ISSUE THREAD

**Issue by raghumirajkar** [awslabs/aws-dotnet-messaging#90]
Hi,

We have a use case where one of the microservices wishes to publish to a SNS topic using the topic ARN (and a few more details like Role to assume and the entire message) from an external service response like an SQS message body or an Event Bridge message. The Topic ARN and other details are part of the SQS/EB message. For ease, we can assume we have a REST API which has all these details in the payload - 
```csharp
[ApiController]
public class DemoController { 
  [HttpPost]
  public IActionResult PublishMessage([FromBody] Dictionay<string, object?> payload) {
          var topicARN = // get topic Arn from payload
          var roleARN = // get Role Arn from payload
          var message =  // get message from payload
          ...
         // 1. STS Assume Role with role ARN provided in the payload
         // 2. Prepare SNS message content using message from payload
         // 3. Publish message to topic provided in payload (topicARN)
   }
}
```
How can we achieve this using AWS Messaging SDK? As seen in the example, the Topic ARN is provided when adding the SNS publisher during service bootstrap -
https://github.com/awslabs/aws-dotnet-messaging/blob/main/sampleapps/PublisherAPI/Program.cs#L21

**Comment by madhub** [comment]
This is valid use case where we need to publish message to customer provided **SQS topic ARN** from a micro service . In multi tenanted deployment we need to store  multiple customers topicARN internally as configuration data & publish message to a topic .

**Comment by ashovlin** [comment]
Currently when registering that we'll publish a type of message, we require the destination as well as the optional message type (if different from the name of the `OrderInfo` type).
```
builder.Services.AddAWSMessageBus(bus =>
{
    bus.AddSNSPublisher<OrderInfo>("arn:aws:sns:us-west-2:012345678910:MPF", "orderInfo");
}
```

For a scenario like you describe, is the destination _always_ determined at runtime? Is there anything like a "default" or "fallback" destination? I'm mulling over whether we should still register these types of messages in advance (to collect the optional message type, as well as anything else we expand that to in the future) vs. specifying everything at the point of publishing a single message.

```
builder.Services.AddAWSMessageBus(bus =>
{
   // Should we still require something like this? Or can that just be specified per-message
   
   // We still want to publish OrderInfo objects as "orderInfo" messages, 
   // but the destination will be set per message
    bus.AddSNSPublisher<OrderInfo>("orderInfo");
}
```

**Comment by raghumirajkar** [comment]
yes, the destination is always determined at runtime and currently we do not have a use case where we want to publish to a default destination

**Comment by ashovlin** [comment]
Thanks, I just created PR #93 but have a few more questions related to the service clients:
1. How large is the "pool" of service clients you use to send these messages? A
2 Are they something you are initializing once and maintaining across different requests/messages? Or are they created for each message?

Currently the PR has something like this, but we were debating if the library should help users track the set of clients to promote their reuse. 
```
await messagePublisher.PublishAsync(message, new SQSOptions
{
    OverrideClient = <client for this message>
});

**Comment by raghumirajkar** [comment]
> Thanks, I just created PR #93 but have a few more questions related to the service clients:
> 
> 1. How large is the "pool" of service clients you use to send these messages? A

The number of clients are unknown ahead of time because while on-boarding to the service, tenants can provide the destination ARN and role to be assumed to publish the message. This information will be stored in the backend which will be used to trigger notification in the flow later

> 2. Are they something you are initializing once and maintaining across different requests/messages? Or are they created for each message?

We're not initializing once and maintaining any data across requests

> 
> Currently the PR has something like this, but we were debating if the library should help users track the set of clients to promote their reuse.
> 
> ```
> await messagePublisher.PublishAsync(message, new SQSOptions
> {
>     OverrideClient = <client for this message>
> });
> ```

**Comment by ashovlin** [comment]
We released 0.2.0-beta today, which included #93. Now you can specify the destination and an override client (with appropriate credentials) at the point of publishing.

```
await _sqsPublisher.SendAsync(message, new SQSOptions
{
    QueueUrl = <override URL>,
    OverrideClient = <override SQS client>
});
```

Let us know how that works for your use case.
